### PR TITLE
[SPARK-26713][CORE][followup] revert the partial fix in ShuffleBlockFetcherIterator

### PR DESCRIPTION
**Pull request by rshkv**
This is a clean cherry-pick of https://github.com/apache/spark/pull/25825, back-ported to our 2.x branch.

We're instances instances recently of task threads hanging while waiting for shuffle results. It seems related to [SPARK-26713](https://issues.apache.org/jira/browse/SPARK-26713). It also matches the difference we have with upstream's latest 2.4.x release.

> There is even a race condition with ShuffledRDD + PipedRDD: the ShuffleBlockFetchIterator is cleaned up at task completion and hangs stdin writer thread, which leaks memory. 

* A partial fix came in here (and we have it): https://github.com/apache/spark/pull/23638
* But IIUC it might have introduced a race condition causing hangs which was later corrected here (this is what we're getting in this PR): https://github.com/apache/spark/pull/25049 

It was easier to revert the commit that introduced the first fix, and then cherry-pick the combined back-port from here: https://github.com/apache/spark/pull/25825 (as opposed to just take the correction).

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala [hunk 1]
```diff
@@ -387,7 +380,7 @@ final class ShuffleBlockFetcherIterator(
     logDebug("Got local blocks in " + Utils.getUsedTimeMs(startTime))
   }
 
-  override def hasNext: Boolean = !isZombie && (numBlocksProcessed < numBlocksToFetch)
+  override def hasNext: Boolean = numBlocksProcessed < numBlocksToFetch
```
Comment (rshkv) [inline core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:383]: IIUC, that `isZombie` flag can cause a race condition https://github.com/apache/spark/pull/23638/files#r251510560

File: core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala [hunk 2]
```diff
@@ -410,7 +403,7 @@ final class ShuffleBlockFetcherIterator(
     // then fetch it one more time if it's corrupt, throw FailureFetchResult if the second fetch
     // is also corrupt, so the previous stage could be retried.
     // For local shuffle block, throw FailureFetchResult for the first IOException.
-    while (!isZombie && result == null) {
+    while (result == null) {
```
Comment (rshkv) [inline core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:406]: IIUC [this thread](https://github.com/apache/spark/pull/23638/files#r251510560) correctly, we can get race conditions on the `isZombie` flag.

File: core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala [hunk 3]
```diff
@@ -410,7 +403,7 @@ final class ShuffleBlockFetcherIterator(
...
       val startFetchWait = System.currentTimeMillis()
       result = results.take()
```
Comment (rshkv) [inline core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:408]: This is where we hang.

**Review by LorenzoMartini — APPROVED** [review]

**Comment by rshkv** [comment]
Thank you, @LorenzoMartini.

**Pull request merged**